Adding a LearnerAnswerDetailsModel

anubhavsinha98 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Adding LearnerAnswerDetailsModel in storage/statistics/gae_models.py to store the responses received from the learner for an answer.
The finalized doc of the model design: [LearnerAnswerDetailsModelDoc](https://docs.google.com/document/d/1clETc5ATpUvHYPWhXjYnZVBsvf4Yz6BHmjoBrK1o1CY/edit?usp=sharing)
## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [ ] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [ ] The PR is made from a branch that's **not** called "develop".
- [ ] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [ ] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [ ] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [ ] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

Commit:
First

Commit:
First

Commit:
Second

Commit:
Third

Commit:
Fourth

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/anubhavsinha98/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Reset

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Added pl json

anubhavsinha98 committed:
Added LearnerAnswerDetailModel

aks681 commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1134,25 @@ def delete_playthroughs_multi(cls, playthrough_ids):
 
 
+class LearnerAnswerDetailModel(base_models.BaseModel):
+    """Model for storing the approach/response user enters when he
```
Maybe change to "Model for storing the approach/response the users enter when they are asked for it" ?

anubhavsinha98 replied:
Okay

anubhavsinha98 replied:
Done

aks681 commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1134,25 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+    # object if it is added in future.
+    entity_type = ndb.StringProperty(required=True, indexed=True)
+    # ID of the response.
+    response_id = ndb.StringProperty(required=True, indexed=False)
```
Why isn't this indexed? What is this used for?

anubhavsinha98 replied:
Yeah I made it true. This can be used to uniquely identify any response.

aks681 reviewed (commented):
Some minor suggestions. 
Also, can you revert the changes to package-lock.json? That is not supposed to happen and is a bug.

anubhavsinha98 committed:
edited docstring

anubhavsinha98 committed:
Modified models and added few functions

anubhavsinha98 commented:
Hi @aks681 I have modified a model a bit and added few functions PTAL. Thanks :)

anubhavsinha98 committed:
Added more functions

aks681 commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1138,115 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+    is_resolved = ndb.BooleanProperty(default=False, required=True)
+
+    @classmethod
+    def get_entity_id(cls, exploration_id, state_name):
```
This function is only for an exploration, right? For any other entities, it would be different. So, maybe rename it to `get_exploration_entity_id` or something similar?

anubhavsinha98 replied:
Yeah it is for exploration only.

aks681 commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1138,115 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+
+    @classmethod
+    def create(
+            cls, exp_id, state_name, entity_type, learner_answer,
```
Ditto here, exp_id and state_name parameters are exclusive to an exploration. Just pass entity_id here and the calling function should take care of passing in the correct entity_id.

anubhavsinha98 replied:
Done

aks681 commented on feconf.py:
```diff
@@ -112,6 +112,7 @@
     TRAINING_JOB_STATUS_FAILED: [TRAINING_JOB_STATUS_NEW]
 }
 
+ENTITY_TYPE_STATE = 'exploration_state'
```
Why is both this and exploration there? Also, is this needed? Entity_type would either be explorations or questions and not states, right?

anubhavsinha98 replied:
Oh my bad, I will remove this.

anubhavsinha98 replied:
Done

aks681 commented on scripts/pre_push_hook.py:
```diff
@@ -340,13 +340,13 @@ def main():
                 continue
             if files_to_lint:
-                lint_status = _start_linter(files_to_lint)
+                lint_status = 0
```
Why are there changed to this file? Was it accidental?

anubhavsinha98 replied:
No I did it to push the things fast, will revert it back once the PR is finalized.

aks681 reviewed (commented):
Also, revert changes to package-lock.json.

anubhavsinha98 committed:
Added tests

anubhavsinha98 committed:
Fixed grammar mistake

anubhavsinha98 commented:
Hi @aks681 PTAL. Thanks!

aks681 reviewed (approved):
LGTM from model perspective. Can you revert the changes to package-lock.json and pre_push_hook.py now?
Also, it seems coverage is reducing. So, maybe add tests for the other untested functions as well?

anubhavsinha98 committed:
pre_push_hook reverted

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' into add_model

anubhavsinha98 committed:
Added test for exception

anubhavsinha98 committed:
Fix lint issues

anubhavsinha98 commented:
Hi @aks681, I have reverted the changes with the other files and also added a test for the remaining exception.
Also @seanlip PTAL. Thanks!

anubhavsinha98 commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1138,132 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+                utils.base64_from_int(utils.get_random_int(_RAND_RANGE)))
+            if not cls.get_by_id(answer_details_id):
+                return answer_details_id
+        raise Exception(
```
Also @aks681, only this line is remaining, as it can't be covered.

aks681 replied:
Ah, ok. This line is an extreme case anyway. It's fine then.

seanlip replied:
It can be covered; talk to @lilithxxx for techniques (or see his recent PRs).

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -63,6 +63,10 @@
     'CyclicStateTransitions': 'state_names'
 }
 
+# Constants used for generating new ids.
```
I'm pretty sure this is copied from somewhere. Try not to copy or define the same thing in multiple places, and reduce duplication where you see it.

anubhavsinha98 replied:
Okay @seanlip, will do this

seanlip replied:
Thanks!

Just a note: I suggest you leave comments inline using "Start a review" (from the "Files Changed" tab), marking comments "Done" as you resolve them, then pushing the updated code and publishing the whole "review" at the end. That way, the code lines up with the review comments and everything is more definite (things are either "done" or "not done", not "will do"), and the reviewer gets notified when there's actually something to review.

This isn't a strong recommendation -- just a suggestion FYI :) Of course, please don't hesitate to start individual threads for anything that you need clarification on prior to resolving!

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1138,132 @@ def delete_playthroughs_multi(cls, playthrough_ids):
 
 
+class LearnerAnswerDetailsModel(base_models.BaseModel):
+    """Model for storing the answer details, learner enters when they
```
... answer details that a learner ...

anubhavsinha98 replied:
Done

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1138,132 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+    """
+    # ID of the entity.
+    entity_id = ndb.StringProperty(required=True, indexed=True)
+    # The type of entity, that whether it is exploration, question, or any
```
The type of entity, e.g. "exploration" or "question".

Also add a "choices" parameter to ndb.StringProperty that restricts the allowable entity types.

seanlip replied:
Please respond to all inline comments before asking for another review.

anubhavsinha98 replied:
Done

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1138,132 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+    # other object if it is added in future.
+    entity_type = ndb.StringProperty(required=True, indexed=True)
+    # The answer submitted by the learner.
+    learner_answer = ndb.StringProperty(required=True, indexed=False)
```
Answers can include lots of different formats. I don't think StringProperty is appropriate here. Probably use JsonProperty but you may run into subsequent issues if the schema changes.

Might also need to store interaction_id here so you know how to "decode" the answer.

Also @aks681 @DubeySandeep Should answers have a schema version, do you think?

aks681 replied:
Hmm, it would be linked to the "answer object" of an interaction, right? So, yeah, it should probably be json property, but could be some existing classes like maybe RuleSpec? (Don't know which class is used to store the answer)

DubeySandeep replied:
Yeah, I also think learner answer should be a JsonProperty and this need to be versioned.

Also, I see similar models, domain objects used in stats (StateAnswers model/object stores metadata and list of SubmittedAnswer object) which are also versioned.

DubeySandeep replied:
I found that we use to store the answer in string format ([here](https://github.com/oppia/oppia/blob/develop/core/domain/stats_domain.py#L986)), I'm not sure whether that's what we are going to save here?

seanlip replied:
I think that loses information; storing the JsonProperty would probably be better.

/cc @brianrodri for info

anubhavsinha98 replied:
Hi @seanlip, as I am using JsonProperty of answers submitted by the learners, I checked that in StateAnswers model we store the submitted_answers, which is actually a domain object which consists of mainly rule_spec_index, answer_group_index, interaction_index, interaction_id and many other fields.
So I can similarly make a domain object which will have attributes as
- answer (submitted by the learner: StringProperty)
- interaction_id
- rule_spec_index
- answer_group_index

which will be necessary to perform queries on the submitted answers. And all these fields will be stored in the JSON in the learner_answer field.
What do you suggest?

seanlip replied:
Don't know -- you'll need to analyze it based on the queries you need to perform to get the data for the frontend. Choose a data structure for the backend to ensure that you can perform each query efficiently. Note that stuff in JSON fields can't be queried, and that get_by_id operations are fastest and should be preferred if possible.

You might want to talk to @aks681 or @DubeySandeep for advice on data modelling. I suggest that you also create a short doc that explains what queries you need to perform (input + expected output), since having all those clearly specified in one place will help with figuring out a good data representation. I'm happy to review that if you like.

anubhavsinha98 replied:
Okay, @seanlip I will make a short doc which will include all the details of the model and share with you.

anubhavsinha98 replied:
Also @seanlip if learner submits any answer it is of string or json property?

seanlip replied:
Treat it as a JSON property, you need to do something generalizable across multiple types of interactions.

anubhavsinha98 replied:
Done

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1138,132 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+    are asked for it.
+
+    The id of instances of this class has the form
+        [ENTITY_TYPE].[ENTITY_ID].[GENERATED_STRING]
```
How are you planning to identify the state/question that this explanation is submitted from?

anubhavsinha98 replied:
This will be differentiated on the basis of entity_type.

seanlip replied:
(See note [here](https://github.com/oppia/oppia/pull/6815#discussion_r290561504) about how entity_type is not an appropriate name for "state".)

anubhavsinha98 replied:
Done

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1138,132 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+    # The answer submitted by the learner.
+    learner_answer = ndb.StringProperty(required=True, indexed=False)
+    # The response submitted by the learner.
+    learner_answer_details = ndb.StringProperty(required=True, indexed=False)
```
ndb.TextProperty (look up the difference)

anubhavsinha98 replied:
Done

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1138,132 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+
+    @classmethod
+    def get_entity_id(cls, id_parameters, entity_type):
+        """Gets entity_id for a batch model based on given entity_type.
```
"batch model"?

anubhavsinha98 replied:
Removed it

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1138,132 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+        """Gets entity_id for a batch model based on given entity_type.
+
+        Args:
+            id_parameters: dict. The parameters which will be
```
No, split this into separate args and document each arg clearly.

anubhavsinha98 replied:
Done

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1138,132 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+        Raises:
+           Exception: If entity type is not valid.
+        """
+        # NOTE: In future if this model is used for any other entity type,
```
I suggest restructuring so that:
- We check if the entity type is valid
- We then construct the ID

Then I don't think you need the comment. Especially if the list used to check for valid entity type is the same as the one used as the "choices" parameter mentioned above.

anubhavsinha98 replied:
Done

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1138,132 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+        return answer_details_instance
+
+    @classmethod
+    def get_answer_details(
```
Shouldn't you be fetching by state? Doesn't really make sense to see all the answer details for different states lumped together.

I'm a bit concerned that this PR suggests that that design consideration hasn't been taken into account; you may want to revisit the pipeline to make sure you're structuring your models for the right sorts of queries in the future. You should also have a plan for extending this to Questions, but I only see Explorations here. (This PR should be designed for that, too, and contain tests for that.)

anubhavsinha98 replied:
It will fetch it by state, as I am using entity_id here which will be in the form 'exploration_id'.'state_name'. So for a particular state, it will get all the answer details.

Also I tried making it generalized as I am using id_parameters:
The id parameters for the exploration will be ```{'exp_id':'the id of exp', 'state_name':'The name of state'}```, if the entity type is a question then there will be different id_parameters like ```{'question_id': 'the id of question', 'state_name': 'the name of the state'}```. There will be a get_id_parameters_for_exp function in the services file which will get the id_parameter for exp, similarly, we can add a function in the services which will generate id_parameter for any entity type like get_id_parameters_for_question. Thus as we will have the entity_type so we can check with if condition which id_parameters function needs to be called. And then we can pass id_parameters with entity_type to generate entity_id for any entity.
Do let me know if you have any doubt in this method?

anubhavsinha98 replied:
![WhatsApp Image 2019-06-05 at 7 35 51 AM](https://user-images.githubusercontent.com/30555544/58925492-56401e00-8765-11e9-9862-c4692105ba99.jpeg)

seanlip replied:
Ah, thanks for explaining. We generally use "entity" to mean a datastore entity, so it's only at the level of "exploration", "question", "skill, etc. but not "state". So I think we need a different naming convention. How do we do it in the generalized review system? /cc @nithusha21 @DubeySandeep 

Also, I think you should keep the ID construction logic in the storage layer, since it is only relevant there and nowhere else (right)? Similar to what we do for some other models like stats. 

I also think you should be implementing specific functions with explicit args for question/exploration in this PR, rather than passing in dicts. Dicts are just data blobs which are hard to validate, and having separate args helps constrain things more and keeps the code more maintainable.

nithusha21 replied:
We call it target_type for 'exploration', 'question', 'topic' etc and target_id for the id (target makes sense in terms of suggestions). 

Maybe you could call this the instance_id?

anubhavsinha98 replied:
Redesigned model doc in the PR explanation.

seanlip reviewed (changes requested):
Thanks! Took a pass.

anubhavsinha98 commented:
Also Sean, should I add the sevices in the same PR or in a different one?

On Wed, Jun 5, 2019, 7:49 AM Sean Lip <notifications@github.com> wrote:

> *@seanlip* commented on this pull request.
> ------------------------------
>
> In core/storage/statistics/gae_models.py
> <https://github.com/oppia/oppia/pull/6815#discussion_r290561504>:
>
> > +                LearnerAnswerDetailsModel.
> +        """
> +        answer_details_id = cls.generate_new_answer_details_id(
> +            entity_type, entity_id)
> +        answer_details_instance = cls(
> +            id=answer_details_id,
> +            entity_id=entity_id,
> +            entity_type=entity_type,
> +            learner_answer=learner_answer,
> +            learner_answer_details=learner_answer_details,
> +            created_on=datetime.datetime.utcnow())
> +        answer_details_instance.put()
> +        return answer_details_instance
> +
> +    @classmethod
> +    def get_answer_details(
>
> Ah, thanks for explaining. We generally use "entity" to mean a datastore
> entity, so it's only at the level of "exploration", "question", "skill,
> etc. but not "state". So I think we need a different naming convention. How
> do we do it in the generalized review system? /cc @nithusha21
> <https://github.com/nithusha21> @DubeySandeep
> <https://github.com/DubeySandeep>
>
> Also, I think you should keep the ID construction logic in the storage
> layer, since it is only relevant there and nowhere else (right)? Similar to
> what we do for some other models like stats.
>
> I also think you should be implementing specific functions with explicit
> args for question/exploration in this PR, rather than passing in dicts.
> Dicts are just data blobs which are hard to validate, and having separate
> args helps constrain things more and keeps the code more maintainable.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/6815?email_source=notifications&email_token=AHJD3GDLFQY74QHWRR62QS3PY4PEXA5CNFSM4HQMPVCKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB2TF2OY#discussion_r290561504>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHJD3GETEBT6ZSVZPUJGJJTPY4PEXANCNFSM4HQMPVCA>
> .
>

seanlip commented:
It's OK for the services to be in a different PR.

anubhavsinha98 committed:
Merge branch 'develop' into add_model

anubhavsinha98 committed:
Fixed backend tests

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Added version

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Made some changes

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Made some changes

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' into add_model

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' into add_model

anubhavsinha98 committed:
Modified model and added domain objects in stats_domain

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' into add_model

anubhavsinha98 commented:
Hi, @seanlip I have changed the design of the model and added domain objects too. I think it needs a fresh review, PTAL Thanks!

seanlip commented:
Could you please add a link to the updated + finalized design doc with all comments fully resolved? That will be necessary context for me reviewing this PR.

Also please respond to all inline comments. You can say "obsolete" if they're obsolete.

Note also the lack of coverage in backend tests, resulting in codecov failures.

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' into add_model

anubhavsinha98 commented:
> Could you please add a link to the updated + finalized design doc with all comments fully resolved? That will be necessary context for me reviewing this PR.
> 
> Also please respond to all inline comments. You can say "obsolete" if they're obsolete.
> 
> Note also the lack of coverage in backend tests, resulting in codecov failures.

Hi @seanlip, I have resolved almost all the comments, I think only one comment is left. Also I have updated the doc as per discussion.  Added the link of the doc in the PR explanation.

Yeah I have not added tests in the PR now, that's why its resulting in the codecov failures.

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1352,115 @@ def validate(self):
+                feconf.CURRENT_LEARNER_ANSWERS_DETAILS_SCHEMA_VERSION)):
+        """Constructs a LearnerAnswerDetail domain object.
+
+            Args:
```
Deindent this and the following lines by 4. (In general, please follow the existing style.)

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1352,115 @@ def validate(self):
+        """Constructs a LearnerAnswerDetail domain object.
+
+            Args:
+                state_reference: str. The reference to the state.
```
I have mentioned this before -- this sort of docstring description is not useful. Please avoid just restating the name of the variable/argument/property in your description, and be more explicit/descriptive about what the possible values are and what they represent. Developers won't know what a "state reference" is, in general.

Please bear this in mind in the future.

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1352,115 @@ def validate(self):
+
+            Args:
+                state_reference: str. The reference to the state.
+                entity_type: str. The type of entity, eg. exploration,
```
You're leaving this too open. Have named constants ENTITY_TYPE_EXPLORATION and ENTITY_TYPE_QUESTION and say that this must be one of [ENTITY_TYPE_EXPLORATION, ENTITY_TYPE_QUESTION].

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1352,115 @@ def validate(self):
+                entity_type: str. The type of entity, eg. exploration,
+                    question etc.
+                interaction_id: str. The ID of the interaction.
+                learner_answer_info_list: list. The list of LearnerAnswerInfo
```
Incomplete typeinfo -- should be of the form `list(???)`. Also I don't understand the "version specified in this object" part, please clarify.

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1352,115 @@ def validate(self):
+                interaction_id: str. The ID of the interaction.
+                learner_answer_info_list: list. The list of LearnerAnswerInfo
+                    objects and version specified in this object.
+                schema_version: str. The schema version of this object.
```
What does "this object" refer to?

Why is it a string?

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1352,115 @@ def validate(self):
+
+    def get_learner_answer_info_dict_list(self):
+        """Returns the learner_answer_info_list stored within this object as
+        a list of LearnerAnswerInfo dicts.
```
Missing Returns part of docstring.

(Please try and do a self-review before requesting a review, to catch issues like this.)

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1352,115 @@ def validate(self):
+        self.learner_answer_info_list = learner_answer_info_list
+        self.schema_version = schema_version
+
+    def get_learner_answer_info_dict_list(self):
```
I don't think you actually need this. Similar to other domain classes, this domain object class should just have a to_dict() method and that's it.

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1352,115 @@ def validate(self):
+    def validate(self):
+        """Validates LearnerAnswerDetails domain object."""
+
+        if not isinstance(self.state_reference, basestring):
```
There is insufficient validation here. Don't just do type validation, do conceptual validation as well. State references must be formatted in a particular way. Entity types can only be one of two things. Interaction ID must correspond to a valid interaction ID in the codebase. All these need to be validated.

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1352,115 @@ def validate(self):
+                'Expected interaction_id to be a string, received %s' % str(
+                    self.interaction_id))
+
+        if not isinstance(self.learner_answer_info_list, list):
```
This is very trivial validation. Validate fully.

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1352,115 @@ def validate(self):
+    """Domain object containing the answer details submitted by the learner."""
+
+    def __init__(self, answer, answer_details, created_on):
+        """Constructs a LearnerAnswerInfo domain object."""
```
Missing docstrings, typeinfo, etc. for the arguments.

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1352,115 @@ def validate(self):
+        )
+
+    def validate(self):
+        """Validates the LearnerAnswerInfo domain object."""
```
Please make the validation fully comprehensive. (See comment above.)

anubhavsinha98 replied:
Done

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -62,6 +62,8 @@
     'CyclicStateTransitions': 'state_names'
 }
+# The choices of entity type valid for LearnerAnswerDetailsModel.
+ENTITY_TYPE_CHOICES = [feconf.ENTITY_TYPE_EXPLORATION]
```
ALLOWED_ENTITY_TYPES

seanlip replied:
Also, what about question? Your PR needs to correctly handle both explorations and questions.

anubhavsinha98 replied:
Done

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1136,224 @@ def delete_playthroughs_multi(cls, playthrough_ids):
 
+class LearnerAnswerDetailsModel(base_models.BaseModel):
+    """Model for storing the answer details that a learner enters when they
+    are asked for it, per state one model instance is created. And currently
```
... are asked for an explanation of their answer. Currently, the model supports exploration states and questions. One instance of this model is created for each of these objects.

The ID of this model is in the format XXX.  <-- you need to specify this too, it's important

anubhavsinha98 replied:
Done

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1136,224 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+    # left for other fields in the model.
+    _MAX_LEARNER_ANSWER_INFO_LIST_BYTE_SIZE = 900000
+
+    # The reference to the state for which model instance is created.
```
Give more info. See above.

anubhavsinha98 replied:
Done

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1136,224 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+    entity_type = ndb.StringProperty(
+        required=True, indexed=True, choices=ENTITY_TYPE_CHOICES)
+    # The id of the interaction for which the answer details were received.
+    interaction_id = ndb.StringProperty(required=True, indexed=False)
```
indexed=True

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1136,224 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+    # The total number of bytes needed to store all of the answers in the
+    # learner_answer_info_list. This value is found by summing the JSON
+    # sizes of all answer info dicts stored inside learner_answer_info_list.
+    accumulated_answer_info_json_size_bytes = ndb.IntegerProperty(
```
It's fine for this to be indexed=True

seanlip replied:
Please respond to **all** review comments

anubhavsinha98 replied:
Done

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1136,224 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+        Returns:
+            str. The state_reference for a new instance of this class.
+        """
+        return '%s.%s' % (exp_id, state_name)
```
Please add tests for state names with unicode chars, spaces, punctuation, etc. (this does happen).

anubhavsinha98 replied:
Done

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1136,224 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+        return question_id
+
+    @classmethod
+    def insert_learner_answer_info_for_exploration(
```
No, you should not do this -- no logic in the storage layer. This should happen in the domain layer. Basically, the domain layer fetches an instance from the storage layer, modifies it, validates it, and puts it. The storage layer should just be a dumb interface that supports get/put/fetch.

(I know there's some logic in the other answers model but that's because that one deals with shards and supplementary storage models. You don't have such a constraint here.)

anubhavsinha98 replied:
Done

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1136,224 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+                new_learner_answer_info)
+
+    @classmethod
+    def insert_learner_answer_info_for_question(
```
Same note as above.

anubhavsinha98 replied:
Done

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1136,224 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+            exp_id: str. ID of the exploration.
+            state_name: str. The name of the state.
+            interaction_id: str. The id of the interaction.
+            learner_answer_info: dict. The dict representation of
```
Here and elsewhere, perhaps you should call it learner_answer_info_dict to distinguish it from the domain object.

anubhavsinha98 replied:
Done

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1136,224 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+    @classmethod
+    def get_model_instance(
+            cls, entity_type, state_reference):
+        """Returns a list of answer details associated with the entity, ordered
```
I have no idea how this docstring relates to the implementation/signature of this method. Please proofread it. It talks about various things (created_on, 'limit' argument) that don't seem to exist in the function.

anubhavsinha98 replied:
Corrected it.

seanlip reviewed (changes requested):
Hi @anubhavsinha98 -- I've taken a pass, thanks.

Please note -- this is a rather weak PR that needs some serious work. There are a lot of proofreading issues, the docstrings are wrong in some places, there are no tests. In general you should write tests with your PRs, rather than waiting for the reviewer to do a review beforehand. Please also do a self-review before submitting to make sure you catch any errors so that the review can go faster, otherwise you'll just get comments like "what is going on here?" and it delays the whole thing.

anubhavsinha98 committed:
Added few changes

anubhavsinha98 committed:
Defined domains and services

anubhavsinha98 committed:
Added tests

anubhavsinha98 committed:
domain tests done

anubhavsinha98 committed:
Fixed lint issues

anubhavsinha98 committed:
Added services test

anubhavsinha98 committed:
Added changes

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' into add_model

anubhavsinha98 committed:
Fixed lint

anubhavsinha98 committed:
Removed unnecessary changes

aks681 commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,360 @@ def validate(self):
+    def __init__(
+            self, state_reference, entity_type, interaction_id,
+            learner_answer_info_list, accumulated_answer_info_json_size_bytes,
+            schema_version=(
```
maybe rename this field to `learner_answer_info_schema_version`? That is what it is, right?

anubhavsinha98 replied:
Yeah it is for the learner answer info, but as it represents the whole model so I think schema_version looks good. What do you suggest?

seanlip replied:
No it doesn't represent the whole model. Schema versions should always represent a **JsonProperty field**. (The only exception is the Exploration schema version which is needed because Explorations also support a YAML representation.)

anubhavsinha98 replied:
Corrected it.

aks681 commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,360 @@ def validate(self):
+                'Expected entity_type to be a string, received %s' % str(
+                    self.entity_type))
+
+        if isinstance(self.entity_type, basestring):
```
This check is not needed as if it weren't, the above validation error would be thrown anyway,

anubhavsinha98 replied:
Done

aks681 commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,360 @@ def validate(self):
+                            self.state_reference))
+            else:
+                raise utils.ValidationError(
+                    'The entity type should be either exploration or question, '
```
Change this to 'Invalid entity type' or similar, as if more entities are added, this error message won't hold.

anubhavsinha98 replied:
Done

aks681 commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,360 @@ def validate(self):
+
+        Args:
+            learner_answer_info: LearnerAnswerInfo. The learner answer info
+                object, which is created after the learner has sbumitted the
```
typo: submitted

anubhavsinha98 replied:
Done

aks681 commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,360 @@ def validate(self):
+                id, which needs to be deleted from
+                the learner_answer_info_list.
+        """
+        deleted_learner_answer_info_id = (
```
Why this extra variable? You can use the argument directly.

anubhavsinha98 replied:
Done

aks681 commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,360 @@ def validate(self):
+        deleted_learner_answer_info_id = (
+            learner_answer_info_id)
+        new_learner_answer_info_list = []
+        for learner_answer_info in self.learner_answer_info_list:
```
Maybe raise an error if such an object is not found?

anubhavsinha98 replied:
Done

aks681 commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,360 @@ def validate(self):
+        Args:
+            learner_answer_info_id: str. The id of the LearnerAnswerInfo object.
+            answer: dict. The answer which is submitted by the learner,
+                actually type of the answer is interaction dependent but mostly
```
learner. Actually the type of answer.....

anubhavsinha98 replied:
Done

aks681 commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,360 @@ def validate(self):
+
+        Args:
+            learner_answer_info_id: str. The id of the LearnerAnswerInfo object.
+            answer: dict. The answer which is submitted by the learner,
```
Change this to dict or str then if it can be string.

anubhavsinha98 replied:
Done

aks681 commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,360 @@ def validate(self):
+    def validate(self):
+        """Validates the LearnerAnswerInfo domain object."""
+
+
```
Remove these blank spaces

anubhavsinha98 replied:
Done

aks681 commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,360 @@ def validate(self):
+                it will be of dict type for most interactions, but it can be
+                string for TextInput interaction.
+            answer_details: str. The details the learner will submit when the
+                learner will be asked questions like 'Hey how did you landed on
```
how did you land on

anubhavsinha98 replied:
Done

aks681 commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,360 @@ def validate(self):
+                string for TextInput interaction.
+            answer_details: str. The details the learner will submit when the
+                learner will be asked questions like 'Hey how did you landed on
+                this answer', 'Why did you picked that answer' etc.
```
Why did you pick

anubhavsinha98 replied:
Done

aks681 commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,360 @@ def validate(self):
+            answer_details: str. The details the learner will submit when the
+                learner will be asked questions like 'Hey how did you landed on
+                this answer', 'Why did you picked that answer' etc.
+            created_on: datetime. The time at which the answer details was
```
details were

aks681 commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,360 @@ def validate(self):
+        if not isinstance(self.id, basestring):
+            raise utils.ValidationError(
+                'Expected id to be a string, received %s' % self.id)
+        if self.answer is None:
```
Maybe check for empty dict also? and empty string? (if it can be a string)

anubhavsinha98 replied:
Done

aks681 commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,360 @@ def validate(self):
+        if sys.getsizeof(self.answer_details) > MAX_ANSWER_DETAILS_BYTE_SIZE:
+            raise utils.ValidationError('The answer details size is to large '
+                                        'to be stored')
+        if not isinstance(self.answer_details, basestring):
```
This check will fail if it's None, right? I'd suggest removing the None check above.

anubhavsinha98 replied:
Done

aks681 commented on core/domain/stats_domain_test.py:
```diff
@@ -1316,3 +1318,342 @@ def test_calculation_output_must_be_less_than_one_million_bytes(self):
+            'exp_id_1.state_name_1')
+
+
+class LearnerAnswerDetailsValidationTests(test_utils.GenericTestBase):
```
Usually all tests for a class (in this case LearnerAnswerDetails) are in the same test class. So, I'd suggest renaming the class above to LearnerAnswerDetailsTests and moving all this into that.

anubhavsinha98 replied:
Done

aks681 commented on core/domain/stats_domain_test.py:
```diff
@@ -1316,3 +1318,342 @@ def test_calculation_output_must_be_less_than_one_million_bytes(self):
+            'Expected accumulated_answer_info_json_size_bytes to be an int')
+
+
+class LearnerAnswerInfoDomainTests(test_utils.GenericTestBase):
```
Ditto about test classes

anubhavsinha98 replied:
Done

aks681 commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1025,198 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
         return None
+
+
+def change_learner_answer_details(
```
maybe rename to apply_change_list?

anubhavsinha98 replied:
As I am not applying list here, as its not the case of exploration where changes will be made first and then applied at the end, in this case, the changes will be automatically applied after performing the action. So apply_change_list might be ambiguous, what do you suggest?

seanlip replied:
I don't actually see why you need a change list and change cmds at all for this -- those are generally used for versioned models where keeping track of the history is important. But that's not the case here, right?

Also, if only one command is possible at a time, have separate functions for each possible operation. Separate controller handlers will map to separate functions. Don't have a mega-function like change_learner_answer_details that puts multiple unrelated things together -- that is not good design.

anubhavsinha98 replied:
Okay I corrected the design now.

aks681 commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1025,198 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+            this will be equal to 'question_id' only.
+        entity_type: str. The type of entity i.e 'exploration' or 'question '.
+        interaction_id: str. The ID of the interaction.
+        change: LearnerAnswerDetailsChange. The change to be applied
```
Oh, is it a single change? I thought it would be a changelist and this function would iterate through those and apply all of them. Have it like that since that would be more general.

anubhavsinha98 replied:
Yeah its a single change.

aks681 commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1025,198 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+    """Returns a LearnerAnswerDetails domain object, with given
+    state_reference, entity_type & interaction_id. This function checks
+    in the datastore if the corresponding LearnerAnswerDetailsModel exists,
+    if not then new domain object is created.
```
This is not done, right? If not present, None is returned.

anubhavsinha98 replied:
Done

aks681 commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1025,198 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+            learner_answer_info_id, change.answer,
+            change.answer_details, datetime.datetime.utcnow())
+        learner_answer_details.add_learner_answer_info(learner_answer_info)
+    elif change.cmd == stats_domain.CMD_DELETE_LEARNER_ANSWER_INFO:
```
Here, if learner_answer_details is None, throw InvalidInputError

anubhavsinha98 replied:
Done

aks681 commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1025,198 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+    elif change.cmd == stats_domain.CMD_DELETE_LEARNER_ANSWER_INFO:
+        learner_answer_details.delete_learner_answer_info(
+            change.learner_answer_info_id)
+    elif change.cmd == stats_domain.CMD_UPDATE_STATE_REFERENCE:
```
Ditto

anubhavsinha98 replied:
Done

aks681 commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1138,122 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+                form 'question_id'.
+
+        Returns:
+            list(LearnerAnswerDetailsModel). List of answer details associated
```
Since we are doing 1 model per state, it won't be a list anymore, right?

anubhavsinha98 replied:
Done

aks681 commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1025,198 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+            entity_type, state_reference))
+    if learner_answer_details_model:
+        learner_answer_details = get_learner_answer_details_from_model(
+            learner_answer_details_model[0])
```
Refer comment about returning lists.

anubhavsinha98 replied:
Done

aks681 commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1025,198 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+        entity_type: str. The type of entity i.e 'exploration' or 'question '.
+    """
+    learner_answer_details.validate()
+    learner_answer_details_model_list = (
```
Ditto

anubhavsinha98 replied:
Done

aks681 commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1138,122 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+                with the entity. Doesn't include deleted entries.
+        """
+        return cls.get_all().filter(cls.entity_type == entity_type).filter(
+            cls.state_reference == state_reference).fetch()
```
If this by default returns a list with only 1 element, extract that element with [0] here itself and return only that.

anubhavsinha98 replied:
Done

aks681 reviewed (changes requested):
I've taken a detailed pass and the PR looks much better now!
Most of my comments are minor nits.

anubhavsinha98 committed:
Addressed comments and added more tests

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,366 @@ def validate(self):
+
+        Args:
+            state_reference: str. This field is used to refer to a state
+                in an exploration or question. Like for an exploration the
```
Like for --> For

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,366 @@ def validate(self):
+                value will be equal to 'exp_id.state_name' & for question
+                this will be equal to 'question_id' only.
+            entity_type: str. The type of entity, for which the domain
+                object is being created, and the value must be one of
```
...created. The value...

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,366 @@ def validate(self):
+
+        split_state_reference = self.state_reference.split('.')
+        if self.entity_type == feconf.ENTITY_TYPE_EXPLORATION:
+            if len(split_state_reference) < 2:
```
Check !=, ditto below. Have your checks be as strict as possible.

anubhavsinha98 replied:
Not equal check can't be used here as state_reference for entity type exploration will have at least one ```.``` in it, so after splitting there should be at least 2 elements in the list.
While in case of question id it's in a hash, therefore, its state reference will not split hence the list will contain only one element.

aks681 replied:
But this is only executed if entity type is exploration, right? In which case it can only be 2, so I'd suggest changing to !=.

anubhavsinha98 replied:
Like for state name which have punctuation in it like 'intro.state' in this case, it can be more than 2.

seanlip replied:
Find a way to construct your reference so that it does not have any characters that can be used in a state name. Perhaps use colon as delimiter (we do that elsewhere)? Or basically find a character that is forbidden in state names and use that.

anubhavsinha98 replied:
Okay, I will look into it.

anubhavsinha98 replied:
I found a similar case of unicode chars state name in the classifier gae_models, although I checked delimiters in the code base, but the mostly ```.``` is used, but somewhere ```:``` & ```-``` is also used. Although in classifier gae_model instance it holds almost same type of instance id of the form ```{{exp_id}}.{{exp_version}}.{{utf8_encoded_state_name}}```, so I think I can go for ```.```

seanlip replied:
Which characters are allowed in state names? I thought we had a regex for that. Please include a link to that code in your next comment (the code that validates what characters are allowed in state names) and use, for your delimiter, some character that is not in that charset.

seanlip replied:
In other words, do what is needed so that you can have a "!= 2" here (unless you can prove that that's impossible).

anubhavsinha98 replied:
https://github.com/oppia/oppia/blob/e69b57106e42f5b7f80e8556b3bd62713f4fadf2/utils.py#L575
https://github.com/oppia/oppia/blob/e69b57106e42f5b7f80e8556b3bd62713f4fadf2/assets/constants.js#L517

I am using ```:``` as a delimilter. I checked in the corresponding lines of the code and found that colon can be used.

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,366 @@ def validate(self):
+            raise utils.ValidationError(
+                'Invalid entity type received %s' % (self.entity_type))
+
+        if not isinstance(self.interaction_id, basestring):
```
You need to also validate that this interaction ID is valid (i.e. corresponds to an actual interaction ID). Validate **everything**.

anubhavsinha98 replied:
Validated.

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,366 @@ def validate(self):
+        Args:
+            learner_answer_info_id: str. The id of the LearnerAnswerInfo object.
+            answer: dict or str. The answer which is submitted by the learner.
+                Actually type of the answer is interaction dependent but mostly
```
This statement is false. Numeric input answers are numbers, for example; other interactions have list types, etc. 

It feels like you have not researched this enough. Please replace this description with a true one.

anubhavsinha98 replied:
Corrected it.

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,366 @@ def validate(self):
+            'id': self.id,
+            'answer': self.answer,
+            'answer_details': self.answer_details,
+            'created_on': str(self.created_on)
```
Convert the datetime to a string using a specific format, don't just cast. Ensure that you have a test checking that from_dict(to_dict(X)) = X.

anubhavsinha98 replied:
Done and added the test for it too.

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,366 @@ def validate(self):
+
+        Args:
+            state_reference: str. This is used to refer to a state
+                in an exploration or question. Like for an exploration the
```
question. For an exploration, the ...

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,366 @@ def validate(self):
+                in an exploration or question. Like for an exploration the
+                value will be equal to 'exp_id.state_name' & for question
+                this will be equal to 'question_id' only.
+            entity_type: str. The type of entity i.e 'exploration' or
```
Refer to the named constants.

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,366 @@ def validate(self):
+        Return:
+            learner_answer_info_id: str. The id generated by the function.
+        """
+        learner_answer_info_id = (state_reference + '.' + entity_type + '.' +
```
We prefer interpolation in this codebase: do '%s.%s.%s%s' % (...)

I don't understand why you need the random strings at the end? Doesn't the entity type and state reference uniquely identify the instance? If not, it probably should.

Also it makes more sense for the entity_type to come before the state_reference. You need to know the former in order to parse the latter correctly.

anubhavsinha98 replied:
Actually, this is the id of each learner answer info which is added in the list. So it needs to have random strings as there can be many learner answer info's, also it should be unique to identify it as it will help in deleting a particular learner answer info.

seanlip replied:
I see, ok. But in that case why do you need state reference and entity type in the ID? It just needs to be some randomly-generated string, right?

anubhavsinha98 replied:
Okay yeah, will correct it.

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,366 @@ def validate(self):
+        if not isinstance(self.answer_details, basestring):
+            raise utils.ValidationError(
+                'Expected answer_details to be a string, received %s' % str(
+                    self.answer_details))
```
Add typeinfo of whatever was actually received (use `type(...)`)

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,366 @@ def validate(self):
+            if self.answer == '':
+                raise utils.ValidationError(
+                    'The answer submitted cannot be an empty string')
+        if not isinstance(self.answer_details, basestring):
```
We should check that this isn't empty, right? (Otherwise why store it? It's not helpful.)

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,205 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+
+    Args:
+        state_reference: str. This is used to refer to a state
+            in an exploration or question. Like for an exploration the
```
See comments above re docstring, they apply here too.

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,205 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+            in an exploration or question. Like for an exploration the
+            value will be equal to 'exp_id.state_name' & for question
+            this will be equal to 'question_id' only.
+        entity_type: str. The type of entity i.e 'exploration' or 'question '.
```
Use named constants in docstring.

anubhavsinha98 replied:
Done

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1138,128 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+    are asked for an explanation of their answer. Currently, the model supports
+    exploration states and questions. One instance of this model is created for
+    each of these objects.
+    The id of this model will be the hash which is automatically generated when
```
No. See how other classes specify the ID and do something similar here. It is really important to specify the generation mechanism clearly.

anubhavsinha98 replied:
Done

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1138,128 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+    # The version of the learner_answer_info_list currently supported by Oppia.
+    # If the internal JSON structure of learner_answer_info_list changes,
+    # CURRENT_SCHEMA_VERSION in this class needs to be incremented.
+    schema_version = ndb.IntegerProperty(
```
Give this a longer name that reflects what it is the schema version of.

anubhavsinha98 replied:
Done

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1138,128 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+        Args:
+            state_reference: str. The reference to the state for which model
+                instance is being created. For exploration state it will be of
+                the form 'exp_id.state_name', & for question it will be of the
```
Do not use "&" in docstrings. Write "and". 

Ditto elsewhere.

anubhavsinha98 replied:
Done

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1138,128 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+            learner_answer_info_list: list. The list of LearnerAnswerInfo
+                objects in dict format, which is defined in the stats_domain.
+            schema_version: int. The version of LearnerAnswerInfo dict, which
+                is currently supported by the Oppia.
```
No, this should be the schema version of the LearnerAnswerInfo dict. It might be older than the "current" schema version if we change the structure in the future.

anubhavsinha98 replied:
Done

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1138,128 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+                with the given entity type and state reference. Doesn't include
+                deleted entries.
+        """
+        model_instance_list = cls.get_all().filter(
```
Ideally this is a get-by-ID operation. You should be able to do that if you actually construct the model IDs deterministically, so that you can reconstruct them here when needed and retrieve the correct model.

anubhavsinha98 replied:
Okay, I was looking to how to generate the ID of the model. As we want to reconstruct it again so we can't use random things. I thought of to use ```entity_type.state_reference```, but it would fail as:
Case1:
exp_id = 'exp.id'
state_name = 'state'
id = 'exploration.exp.id.state'

Case2:
exp_id = 'exp'
state_name = 'id.state'
id = 'exploration.exp.id.state'

So in this case it will fail, any suggestions?

anubhavsinha98 replied:
Got the idea, I will use the hashlib library.

anubhavsinha98 replied:
But yeah for hash I need to specifically pass the exp_id and state_name (question_id for question).
What do you suggest @seanlip and @aks681 ?

aks681 replied:
I don't think the cases you told above would occur, as exp id itself is a random hash. So, it might be fine to just use exp id and state names.

anubhavsinha98 replied:
Oh, then it's fine. Thanks :)

anubhavsinha98 replied:
Done

seanlip replied:
This is good thinking, and you should use a delimiter that is not used in the charset of either the exp ID or the state names when constructing the composite ID, so that the names can be parsed uniquely. You should not need to encode either of these components though using hashlib, though make sure you consider the case of unicode state names and include tests for them (it's possible that those might need utf-8 encoding, I am not sure).

See what works with GAE -- the less encoding the better, but it does need to actually work.

anubhavsinha98 replied:
I think storing can be done with ```.``` delimiter as exp_id has its own hash value so the model id will be unique. Okay I will test it for unicode state name and check if it can find the model instance.

seanlip reviewed (changes requested):
Took a pass -- looks better, but still needs work. PTAL.

Also, if you want to, it's fine to add the controller layer, but this is optional. You might want to do it in the interests of time, though.

anubhavsinha98 committed:
Addressed changes

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' into add_model

anubhavsinha98 committed:
Changed schema version

anubhavsinha98 committed:
Corrected doc string and id function in learner answer info

aks681 commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,321 @@ def validate(self):
+                    'should be of the form \'exp_id.state_name\', but '
+                    'received %s' % (self.state_reference))
+        elif self.entity_type == feconf.ENTITY_TYPE_QUESTION:
+            if len(split_state_reference) > 1:
```
Ditto, != 1, as this is only executed if entity type is question.

anubhavsinha98 replied:
Okay yeah for question it can be done

anubhavsinha98 replied:
Done

aks681 commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,239 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+        learner_answer_details = get_learner_answer_details_from_model(
+            learner_answer_details_model)
+        return learner_answer_details
+    else:
```
This else isn't needed, if the above condition is executed, because of the return, it wouldn't come here anyway.

anubhavsinha98 replied:
Done

aks681 commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,239 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+        learner_answer_details.entity_type,
+        learner_answer_details.state_reference)
+    if learner_answer_details_model:
+        if learner_answer_details_model.id == instance_id:
```
Just checking, is this for the case where the creator has renamed a state?

anubhavsinha98 replied:
Yeah its when the creator renamed the state.

aks681 commented on core/storage/statistics/gae_models.py:
```diff
@@ -1134,6 +1138,149 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+        model_instance = cls.get(instance_id, strict=False)
+        if model_instance:
+            return model_instance
+        else:
```
Ditto about not needing else

anubhavsinha98 replied:
Done

aks681 reviewed (commented):
Looks good!
Also, revert changes to package-lock.json

seanlip commented:
@anubhavsinha98 please make sure to go back through the previous comments and address **every one**. Some comments from previous reviews have not been responded to.

anubhavsinha98 committed:
Addressed comments

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' into add_model

anubhavsinha98 committed:
Removed extra variable in feconf

anubhavsinha98 commented:
Hi @seanlip, I have addressed all the comments. PTAL Thanks!

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1450,7 +1395,7 @@ def __init__(
                 learner_answer_info_list in bytes.
-            schema_version: int. The schema version of the
+            learner_answer_info_schema_version: int. The schema version of the
                 LearnerAnswerInfo class.
```
... LearnerAnswerInfo dicts stored in this instance.

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1675,16 +1627,16 @@ def get_learner_answer_info_id(cls, state_reference, entity_type):
             learner_answer_info_id: str. The id generated by the function.
         """
-        learner_answer_info_id = (state_reference + '.' + entity_type + '.' +
+        learner_answer_info_id = (entity_type + '.' + state_reference + '.' +
```
I think in an earlier comment I mentioned that we prefer interpolation.

Also you still need to take out the first two parts.

anubhavsinha98 replied:
Done

seanlip reviewed (commented):
I started doing a review and found that you have not addressed all the comments, so I stopped the review midway. Please check that you have replied to and addressed every single historical comment, and then feel free to ask for a review again. Thanks.

anubhavsinha98 committed:
Changed id function

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' into add_model

anubhavsinha98 committed:
Made few changes

anubhavsinha98 committed:
Added tests for unicode chars

anubhavsinha98 commented:
Hi @seanlip, I have addressed the comments. PTAL Thanks!

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -1351,3 +1365,312 @@ def validate(self):
+        learner_answer_info_dict_size = (
+            learner_answer_info.get_learner_answer_info_dict_size())
+        if (self.accumulated_answer_info_json_size_bytes +
+                learner_answer_info_dict_size <= (
```
We need a test for "big" answers that cause the limit to be exceeded. I don't see such a test in the test file, could you please add it?

anubhavsinha98 replied:
Added

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+    learner_answer_details_model = (
+        stats_models.LearnerAnswerDetailsModel.get_model_instance(
+            entity_type, state_reference))
+    if learner_answer_details_model:
```
if XXX is not None

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+    """Saves the LearnerAnswerDetails domain object in the datatstore.
+
+    Args:
+        learner_answer_details: LearnerAnswerDetails. The  learner answer
```
drop extra space before "learner"

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+
+
+def save_learner_answer_details(
+        learner_answer_details, state_reference, entity_type):
```
Have entity_type come before state_reference (here and in other arglists). Make sure to reorder the Args section accordingly.

In fact, better to order as: entity_type, state_reference, learner_answer_details (provide the address first, then the thing that should be saved there).

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+        learner_answer_details.entity_type,
+        learner_answer_details.state_reference)
+    if learner_answer_details_model:
+        if learner_answer_details_model.id == instance_id:
```
I think you should call out really clearly in the docstring that learner_answer_details_model.entity_type and learner_answer_details_model.state_reference may not match the passed-in entity type and state reference, and describe the cases where this may occur, since that's not really intuitive.

Even better would be to make sure that they do match (and have the callers handle the other cases where deleting/recreating is needed), so that there is no confusion about which entity type or state reference is the "correct" one. In other words, for any function with two state references as part of its args, there should be a concept of "old state reference" and "new state reference".

anubhavsinha98 replied:
Done

anubhavsinha98 replied:
Done, actually I added a function which creates a new instance

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+        learner_answer_details.state_reference)
+    if learner_answer_details_model:
+        if learner_answer_details_model.id == instance_id:
+            learner_answer_details_model.state_reference = (
```
In what situations would state_reference and entity_type need to be set again? I think this can never happen, but perhaps I'm missing something.

anubhavsinha98 replied:
Oh, I think if id and instance id is same, then it need not be set.

anubhavsinha98 replied:
But yeah if they are different then the model needs to be deleted and then newly create it with the new model instance id.

seanlip replied:
But here they are the same right?

In any case, see second paragraph of this comment: https://github.com/oppia/oppia/pull/6815#discussion_r296427158 -- it would be better if you can structure your functions that way, to avoid confusion.

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+            learner_answer_details_model.put()
+        else:
+            learner_answer_details_model.delete()
+            stats_models.LearnerAnswerDetailsModel.create_model_instance(
```
Can you find a way to not duplicate this and lines 1130-1137?

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+
+
+def record_learner_answer_info(
+        state_reference, entity_type, interaction_id, answer, answer_details):
```
Put entity_type before state_reference

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+    """
+    learner_answer_details = get_learner_answer_details(
+        state_reference, entity_type)
+    if not learner_answer_details:
```
if XXX is not None

(use this construct when you are not dealing with implicit truthy/falsy conversion)

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+        learner_answer_details = stats_domain.LearnerAnswerDetails(
+            state_reference, entity_type, interaction_id, [], 0)
+    learner_answer_info_id = (
+        stats_domain.LearnerAnswerInfo.get_learner_answer_info_id())
```
Can we rename the method to get_new_...? Otherwise, unclear that it's creating a new thing.

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+            or ENTITY_TYPE_QUESTION, which are declared in feconf.py.
+        learner_answer_info_id: str. The unique ID of the learner answer info
+            which needs to be deleted.
+"""
```
Indent by 4.

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+"""
+    learner_answer_details = get_learner_answer_details(
+        state_reference, entity_type)
+    if not learner_answer_details:
```
if XXX is None

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+
+def update_state_reference(
+        new_state_reference, old_state_reference, entity_type):
+    """Updates the state reference field of the model instance if the
```
Sentence doesn't scan, please rewrite

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+
+
+def update_state_reference(
+        new_state_reference, old_state_reference, entity_type):
```
entity_type, old_state_reference, new_state_reference

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+            or ENTITY_TYPE_QUESTION, which are declared in feconf.py.
+    """
+    learner_answer_details = get_learner_answer_details(
+        old_state_reference, entity_type)
```
entity_type, old_state_reference

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+    """
+    learner_answer_details = get_learner_answer_details(
+        old_state_reference, entity_type)
+    if not learner_answer_details:
```
if XXX is None

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+    learner_answer_details_model = (
+        stats_models.LearnerAnswerDetailsModel.get_model_instance(
+            feconf.ENTITY_TYPE_EXPLORATION, state_reference))
+    if learner_answer_details_model:
```
if XXX is not None

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+
+def delete_learner_answer_details_for_exploration_state(
+        exp_id, state_name):
+    """Delete the LearnerAnswerDetailsModel instance of entity type
```
You can just say: Deletes the LearnerAnswerDetailsModel corresponding to the given exploration ID and state name.

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+
+
+def delete_learner_answer_details_for_question_state(question_id):
+    """Delete the LearnerAnswerDetailsModel instance of entity type
```
Ditto: Deletes the LearnerAnswerDetailsModel for the given question ID.

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+    learner_answer_details_model = (
+        stats_models.LearnerAnswerDetailsModel.get_model_instance(
+            feconf.ENTITY_TYPE_QUESTION, state_reference))
+    if learner_answer_details_model:
```
if XXX is not None

anubhavsinha98 replied:
Done

seanlip reviewed (changes requested):
This is looking pretty good! Just one more round, I think this should be the last one.

anubhavsinha98 committed:
Addressed reviews and comments

anubhavsinha98 committed:
Merge branch 'develop' into add_model

anubhavsinha98 committed:
Added comments

anubhavsinha98 commented:
Hi @seanlip made the changes. PTAL Thanks!

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1159,76 +1162,76 @@ def record_learner_answer_info(
+        old_state_reference: str. The old state reference which needs to be
+            changed.
+        new_state_reference: str. The new state reference which needs to be
+            updated..
```
Drop extra period at end.

anubhavsinha98 replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -1159,76 +1162,76 @@ def record_learner_answer_info(
     learner_answer_details.update_state_reference(new_state_reference)
     save_learner_answer_details(
-        learner_answer_details, old_state_reference, entity_type)
+        entity_type, old_state_reference, learner_answer_details,)
```
Drop extra comma at end.

anubhavsinha98 replied:
Done

seanlip reviewed (approved):
LGTM. Thanks!

seanlip commented:
@aks681 @brianrodri @vojtechjelinek @nithusha21 PTAL as codeowners

anubhavsinha98 committed:
Merge branch 'develop' into add_model

nithusha21 reviewed (approved):
LGTM as a codeowner. Thanks!

anubhavsinha98 committed:
Fixed , and .

anubhavsinha98 committed:
Update package-lock.json

aks681 commented on feconf.py:
```diff
@@ -198,6 +199,10 @@
 
+# This value should be updated if the schema of LearnerAnswerInfo
+# dict schema changes.
+CURRENT_LEARNER_ANSWERS_INFO_SCHEMA_VERSION = 1
```
Rename this to `CURRENT_LEARNER_ANSWER_INFO_SCHEMA_VERSION`, since the field name is `learner_answer_info_schema_version`.

anubhavsinha98 replied:
Done

aks681 reviewed (approved):
LGTM!

anubhavsinha98 committed:
Changed constant

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' into add_model
